src/main/java/org/carlmontrobotics/subsystems/Limelight.java

```diff
@@ -89,16 +102,59 @@ public double getDistanceToNoteMeters() {
     }
   }
 
-  public double getArmAngleToShootSpeakerRad(){
+  public double getArmAngleToShootSpeakerRad() {
     double armRestingHeightToSubwooferMeters = HEIGHT_FROM_RESTING_ARM_TO_SPEAKER_METERS;
     double horizontalDistanceMeters = getDistanceToSpeakerMeters() + SIDEWAYS_OFFSET_TO_OUTTAKE_MOUTH;
     return END_EFFECTOR_BASE_ANGLE_RADS - Math.atan(armRestingHeightToSubwooferMeters / horizontalDistanceMeters);
   }
 
-  public double getRotateAngleDeg() {
+  public double getRotateAngleRad() {
     double cameraLensHorizontalOffset = getTXDeg(SHOOTER_LL_NAME) / getDistanceToSpeakerMeters();
     double realHorizontalOffset = Math.atan(cameraLensHorizontalOffset / getDistanceToSpeakerMeters());
     return Math.atan(realHorizontalOffset / getDistanceToSpeakerMeters());
   }
 
+  // megatag2
+
+  public void updateMT2Odometry() {
+    boolean rejectVisionUpdate = false;
+
+    LimelightHelpers.SetRobotOrientation(SHOOTER_LL_NAME,
+        poseEstimator.getEstimatedPosition().getRotation().getDegrees(), 0, 0, 0, 0, 0);
+    LimelightHelpers.PoseEstimate visionPoseEstimate = LimelightHelpers
+        .getBotPoseEstimate_wpiBlue_MegaTag2(SHOOTER_LL_NAME);
+
+    if (Math.abs(drivetrain.getGyroRate()) > MAX_TRUSTED_ANG_VEL) {
+      rejectVisionUpdate = true;
+    }
+
+    if (visionPoseEstimate.tagCount == 0) {
+      rejectVisionUpdate = true;
+    }
+
+    if (!rejectVisionUpdate) {
+      poseEstimator
+          .setVisionMeasurementStdDevs(VecBuilder.fill(STD_DEV_X_METERS, STD_DEV_Y_METERS, STD_DEV_HEADING_RADS));
+      poseEstimator.addVisionMeasurement(visionPoseEstimate.pose, visionPoseEstimate.timestampSeconds);
+    }
+  }
+
+  public double getRotateAngleRadMT2() {
+    Pose3d targetPoseRobotSpace = LimelightHelpers.getTargetPose3d_RobotSpace(SHOOTER_LL_NAME); // pose of the target
+
+    double targetX = targetPoseRobotSpace.getX(); // the forward offset between the center of the robot and target
+    double targetY = targetPoseRobotSpace.getY(); // the sideways offset
+
+    double targetOffsetRads = Math.atan2(targetY, targetX);
```

Comment:
double check that the signs are correct (to be clear, idk if they are correct, but people not understanding the coordinate system has caused bugs in the past)

Reply:
what signs?

Reply:
i'll flip signs if necessary after testing 🤷